lychee-lib/src/extract/html5gum.rs

```diff
@@ -0,0 +1,206 @@
+use html5gum::{Emitter, Error, Tokenizer};
+
+use super::plaintext::extract_plaintext;
+use crate::types::raw_uri::RawUri;
+
+#[derive(Clone)]
+struct LinkExtractor {
+    links: Vec<RawUri>,
+    current_string: Vec<u8>,
+    current_tag_name: Vec<u8>,
+    current_tag_is_closing: bool,
+    current_attribute_name: Vec<u8>,
+    current_attribute_value: Vec<u8>,
+    last_start_tag: Vec<u8>,
+}
+
+/// this is the same as `std::str::from_utf8_unchecked`, but with extra debug assertions for ease
+/// of debugging
+unsafe fn from_utf8_unchecked(s: &[u8]) -> &str {
+    debug_assert!(std::str::from_utf8(s).is_ok());
+    std::str::from_utf8_unchecked(s)
+}
+
+impl LinkExtractor {
+    pub(crate) const fn new() -> Self {
```

Comment:
I think we could `#[derive(Default)]` for `LinkExtractor` and this code goes away?

Reply:
I did that now, and it makes `new` non-const as `Default` is not const